[cloud-gce] Move integration tests to unit tests

dadoonet opened the pull request:
Basically this PR:

* Update GCE API to latest version (1.71)
* Replace IT by unit tests
* Use Google Transport Mock to simulate JSON Response which helps to **really** test the implementation classes. That was not the case previously.
* Remove GceSimpleITest as we now have RestIT (same goal for boths)

Would love to get feedback from reviewers. Thanks!

nik9000 commented on plugins/cloud-gce/src/main/java/org/elasticsearch/cloud/gce/GceComputeServiceImpl.java:
```diff
@@ -49,26 +48,26 @@
                         InstanceList instanceList = list.execute();
                         if (instanceList.isEmpty()) {
-                            return new ArrayList();
+                            return Lists.newArrayList();
```
I just saw @s1monw replacing ```Lists.newArrayList``` with ```new ArrayList<>()``` a few days ago. Maybe that way is out preferred way now?

+1 on removing the warning though.

s1monw replied:
yeah please just return Collections.EMPTY_LIST here or so?

dadoonet commented:
Tests are failing with:

```
Started J0 PID(13396@MacBook-Air-de-David.local).
Suite: org.elasticsearch.discovery.gce.GceDiscoveryTest
  2> REPRODUCE WITH: mvn test -Pdev -Dtests.seed=31458D3312839D7E -Dtests.class=org.elasticsearch.discovery.gce.GceDiscoveryTest -Des.logger.level=ERROR -Dtests.assertion.disabled=false -Dtests.security.manager=true -Dtests.heap.size=512m -Dtests.locale=fr_FR -Dtests.timezone=Europe/Paris
  2> NOTE: test params are: codec=Asserting(Lucene50): {}, docValues:{}, sim=RandomSimilarityProvider(queryNorm=false,coord=crazy): {}, locale=vi_VN, timezone=Australia/Sydney
  2> NOTE: Mac OS X 10.10.4 x86_64/Oracle Corporation 1.7.0_60 (64-bit)/cpus=4,threads=1,free=463145336,total=515375104
  2> NOTE: All tests run in this JVM: [GceDiscoveryTest]
ERROR   0.00s | GceDiscoveryTest (suite) <<<
   > Throwable #1: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessClassInPackage.sun.misc")
   >    at __randomizedtesting.SeedInfo.seed([31458D3312839D7E]:0)
   >    at java.security.AccessControlContext.checkPermission(AccessControlContext.java:372)
   >    at java.security.AccessController.checkPermission(AccessController.java:559)
   >    at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
   >    at java.lang.SecurityManager.checkPackageAccess(SecurityManager.java:1525)
   >    at java.lang.Class.checkPackageAccess(Class.java:2304)
   >    at java.lang.Class.checkMemberAccess(Class.java:2284)
   >    at java.lang.Class.getDeclaredFields(Class.java:1805)
   >    at java.lang.Thread.run(Thread.java:745)
Completed [1/2] in 1.78s, 9 tests, 1 error <<< FAILURES!
```

Not sure what I should look at from this? Any idea @rjernst @rmuir ?

nik9000 commented on plugins/cloud-gce/src/main/java/org/elasticsearch/cloud/gce/GceComputeServiceImpl.java:
```diff
@@ -120,27 +126,27 @@ public synchronized Compute client() {
                     .build();
         } catch (Exception e) {
-            logger.warn("unable to start GCE discovery service: {} : {}", e.getClass().getName(), e.getMessage());
+            logger.warn("unable to start GCE discovery service: {} : {}", e, e.getClass().getName(), e.getMessage());
```
This looks not quite right. I'm not 100% sure, but I'd do:
```
+            logger.warn("unable to start GCE discovery service", e);
```
and let the logger do its job with the exception. I'm tired though - airplanes. So don't take what I type at face value.

rmuir commented on core/src/main/resources/org/elasticsearch/bootstrap/security.policy:
```diff
@@ -92,7 +92,9 @@ grant {
   permission java.lang.RuntimePermission "accessClassInPackage.sun.security.ssl";
-  
+  // needed by GCE tests sdk
+  permission java.lang.RuntimePermission "accessClassInPackage.sun.util.calendar";
```
-1: do not add permissions to sun internal classes because tests have static leaks. fix the leaks instead.

dadoonet replied:
I'd love to! Any idea on how I can find which class is responsible for those errors ?

rmuir replied:
Look for statics in any test and null then out in afterclass

dadoonet commented:
Thanks to @rmuir comment, I found my issue (and learned from that BTW) and fixed it. It's now running fine. Also fixed comments left in the first review.

Let me know!

I consider this PR as a first try on GCE. I'd like to do next the same cleanup on azure and aws if possible.

dadoonet commented:
I marked it as a blocker as we have a `AwaitsFix` in it.

s1monw commented on plugins/cloud-gce/src/main/java/org/elasticsearch/cloud/gce/GceComputeServiceImpl.java:
```diff
@@ -96,17 +96,24 @@
         this.project = settings.get(Fields.PROJECT);
         String[] zoneList = settings.getAsArray(Fields.ZONE);
-        this.zoneList = Lists.newArrayList(zoneList);
+        this.zones = Lists.newArrayList(zoneList);
```
please use Arrays.asList while we re here

s1monw commented on plugins/cloud-gce/src/main/java/org/elasticsearch/cloud/gce/GceComputeServiceImpl.java:
```diff
@@ -96,17 +96,24 @@
+        this.zones = Lists.newArrayList(zoneList);
+    }
+
+    protected HttpTransport getGceHttpTransport() throws GeneralSecurityException, IOException {
```
can this be private and synchronized please

s1monw replied:
do we close this transport as well at some point?

s1monw commented on plugins/cloud-gce/src/test/java/org/elasticsearch/discovery/gce/GceComputeServiceMock.java:
```diff
@@ -0,0 +1,122 @@
+        logger.debug("--> read mock file from [{}]", mockFileName);
+        URL resource = GceComputeServiceMock.class.getResource(mockFileName);
+        try (InputStream is = resource.openStream()) {
+            final StringBuffer sb = new StringBuffer();
```
can this be a string builder rather than StringBuffer?

s1monw commented on plugins/cloud-gce/src/test/java/org/elasticsearch/discovery/gce/GceComputeServiceMock.java:
```diff
@@ -0,0 +1,122 @@
+            logger.trace("{}", response);
+            return response;
+        } catch (IOException e) {
+            logger.error("can not read mock definition file [{}]", mockFileName);
```
why don't we throw the IOException up again?

s1monw replied:
why do you log this instead? Can't you throw it up?

dadoonet replied:
I do both. Do you mean remove the log line?

s1monw commented:
left some comments - this looks awesome

dadoonet commented:
@s1monw Updated with your comments.

s1monw commented:
left one comment LGTM otherwise
